File: leetcode/0430-flatten-a-multilevel-doubly-linked-list/solution.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* prev;
    Node* next;
    Node* child;
};
*/

class Solution {
public:
    /**
     * RUNTIME: 4 ms (81.75%)
     * MEMORY: 7.3 MB (63.12%)
    */
    Node* flatten(Node* head) {
        Node *curr;
        curr = head;
        while (curr != NULL) {
            if (curr->child != NULL) {
                getChild(curr, curr->child, curr->next);
                curr->child = NULL;
            }
            curr = curr->next;
        }
        return head;
    }

    void getChild(Node* before, Node* child, Node* after) {
        // bind before to child
        before->next = child;
        // bind child to before
        child->prev = before;
        // bind child's lastNode to after
        Node *lastNode;
        lastNode = child;
        while (child != NULL) {
            if (child->child != NULL) {
                getChild(child, child->child, child->next);
                child->child = NULL;
            }
            lastNode = child;
            child = child->next;
        }
        lastNode->next = after;
        // bind after to child's lastNode
        if (after != NULL) {
            after->prev = lastNode;
        }
    }
};

File: leetcode/0430-flatten-a-multilevel-doubly-linked-list/solution.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, prev, next, child):
        self.val = val
        self.prev = prev
        self.next = next
        self.child = child
"""

class Solution:
    def flatten(self, head: 'Optional[Node]') -> 'Optional[Node]':
        """
        RUNTIME: 35 ms (93.75%)
        MEMORY: 14.9 MB (27.36%)
        """
        curr = head
        while curr is not None:
            if curr.child is not None:
                self.get_child(curr, curr.child, curr.next)
                curr.child = None
            curr = curr.next
        return head

    def get_child(self, before, child, after):
        # bind prev to child
        before.next = child
        # bind child to prev
        child.prev = before
        # bind child's last_node to after
        last_node = child
        while child is not None:
            if child.child is not None:
                self.get_child(child, child.child, child.next)
                child.child = None
            last_node = child
            child = child.next
        last_node.next = after
        # bind after to child's last_node
        if after is not None:
            after.prev = last_node
